### ISSUE TITLE
backpropagation

### ISSUE THREAD

**Issue by YQ0Sun** [ultralytics/ultralytics#17385]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, I would like to find the code for backpropagation to understand how the model parameters are optimized after the loss function. I look forward to your reply.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @YQ0Sun, thank you for your interest in Ultralytics 🚀! We encourage new users to explore our [Docs](https://docs.ultralytics.com) where you'll find a wealth of [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) guides and examples which may already answer your question.

If this is a ❓ Question regarding backpropagation, please refer to our documentation on how the optimization process is implemented. Additionally, providing context, such as specific sections of code you're analyzing, can be helpful.

If this is a 🐛 Bug Report or unexpected behavior you're observing, we kindly ask you to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to assist us in diagnosing the issue effectively.

Feel free to engage with our vibrant community in places like our [Discord](https://ultralytics.com/discord) 🎧, [Discourse](https://community.ultralytics.com), or the [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure that you are using the latest version of `ultralytics` with all necessary [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python≥3.8**](https://www.python.org/) environment with [**PyTorch≥1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

Run YOLO in any of these recommended environments, which have all dependencies preinstalled, including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

A green badge means all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. CI tests ensure that all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) function correctly on macOS, Windows, and Ubuntu every day and with every commit.

This is an automated response, but rest assured, an Ultralytics engineer will review your query soon. 😊

**Comment by Y-T-G** [comment]
https://github.com/ultralytics/ultralytics/blob/3c976807b878af320e5bd85c6147bd321db26daa/ultralytics/engine/trainer.py

**Comment by YQ0Sun** [comment]
@Y-T-G 
First, thank you for your response.
`   # Backward
                self.scaler.scale(self.loss).backward()`
I found the trainer.py function, and I located Backward. I’d like to know if it’s this line: "self.scaler.scale(self.loss).backward()"? It’s in the _do_train function within the BaseTrainer class, right?

`def backward(
        self, gradient=None, retain_graph=None, create_graph=False, inputs=None
    ):
        r"""Computes the gradient of current tensor wrt graph leaves.

        The graph is differentiated using the chain rule. If the tensor is
        non-scalar (i.e. its data has more than one element) and requires
        gradient, the function additionally requires specifying ``gradient``.
        It should be a tensor of matching type and location, that contains
        the gradient of the differentiated function w.r.t. ``self``.

        This function accumulates gradients in the leaves - you might need to zero
        ``.grad`` attributes or set them to ``None`` before calling it.
        See :ref:`Default gradient layouts<default-grad-layouts>`
        for details on the memory layout of accumulated gradients.

        .. note::

            If you run any forward ops, create ``gradient``, and/or call ``backward``
            in a user-specified CUDA stream context, see
            :ref:`Stream semantics of backward passes<bwd-cuda-stream-semantics>`.

        .. note::

            When ``inputs`` are provided and a given input is not a leaf,
            the current implementation will call its grad_fn (though it is not strictly needed to get this gradients).
            It is an implementation detail on which the user should not rely.
            See https://github.com/pytorch/pytorch/pull/60521#issuecomment-867061780 for more details.

        Args:
            gradient (Tensor or None): Gradient w.r.t. the
                tensor. If it is a tensor, it will be automatically converted
                to a Tensor that does not require grad unless ``create_graph`` is True.
                None values can be specified for scalar Tensors or ones that
                don't require grad. If a None value would be acceptable then
                this argument is optional.
            retain_graph (bool, optional): If ``False``, the graph used to compute
                the grads will be freed. Note that in nearly all cases setting
                this option to True is not needed and often can be worked around
                in a much more efficient way. Defaults to the value of
                ``create_graph``.
            create_graph (bool, optional): If ``True``, graph of the derivative will
                be constructed, allowing to compute higher order derivative
                products. Defaults to ``False``.
            inputs (sequence of Tensor): Inputs w.r.t. which the gradient will be
                accumulated into ``.grad``. All other Tensors will be ignored. If not
                provided, the gradient is accumulated into all the leaf Tensors that were
                used to compute the attr::tensors.
        """
        if has_torch_function_unary(self):
            return handle_torch_function(
                Tensor.backward,
                (self,),
                self,
                gradient=gradient,
                retain_graph=retain_graph,
                create_graph=create_graph,
                inputs=inputs,
            )
        torch.autograd.backward(
            self, gradient, retain_graph, create_graph, inputs=inputs
        )`
  I clicked into the `backward()` function. Is this the main part of backpropagation optimization, right?

> 

>

**Comment by YQ0Sun** [comment]
# Backward 
   self.scaler.scale(self.loss).backward()
   "So is the backpropagation optimization done here: `self.scaler.scale(self.loss).backward()`? I found it in the `trainer.py` function."

**Comment by Y-T-G** [comment]
Yes

**Comment by YQ0Sun** [comment]
ok，thanks @Y-T-G
